connectivity_matrix code speed up

DelinteNicolas opened the pull request:
Following discussion #2972 , here is a pull request with the new code. The main objective was to speed up the code when 'inclusive' was set to True.

For a tractogram of 500 000 streamlines, the code now runs in less than 20s with all parameters set to True, compared to more than 2 minutes previously. The code is also shorter and subjectively more readable.

Main differences:
- 'inclusive' is now faster, with a x3 speed up for a tractogram of 2000 streamlines and >x6 speed up for a tractogram of 500 000 streamlines. Speed when 'inclusive' is set to False is unchanged.
- 'symmetric' now returns only the top half of the matrix when set to False. The results thus differ from the previous implementation when set to False. To be honest, 'symmetric'=False does not make sense to me since streamlines are direction agnostic. Feel free to change this behavior.
- 'return_mapping' is identical
- 'mapping_as_streamlines' was slightly changed. The results are identical but the function now returns a 'mapping_as_streamlines' map even when 'return_mapping' is set to False. Feel free to also change this behavior. It would make sense to me to change the parameter name to 'return_mapping_as_streamlines' to make the behavior more explicit. 

The results were only tested with a few examples, more testing should be performed to ensure reproducibility between both implementations.

skoudoro commented:
Thank you for this @DelinteNicolas!  Not sure yet why the CI's are failing. it seems that your news implementation does not provide identical results as the previous implementation since tests are failing. 

I will look why a bit later.

Thanks !

DelinteNicolas commented:
Hello, I have modified the pull request to simplify the changes you would want to make. The implementation now has the same behavior as the previous implementation for the 'symmetric' and 'mapping_as_streamlines' parameters.

Main changes:
- changed matrix type to int64 to fit with previous implementation
- removed np.floor and astype(int) which should be handled by _to_voxel_coordinates function
- changed behavior of mapping_as_streamlines to work similarly the previous implementation
- inclusive default set to False
- symmetry matrix has been changed to fit with previous behavior. It no longer returns the top half of the symmetric matrix
- mapping incorrectly returned streamlines endpoints when mapping_as_streamlines as True. This has been fixed

The tests should now be correct but **/!\ results still differ when inclusive is set to True and symmetric to False**

Hope this helps

skoudoro reviewed (approved):
LGTM. Thank you @DelinteNicolas. 

I will wait until Tuesday if there is another reviewer. If not, I will go ahead and merge it.

arokem commented:
This looks fine to me. I just noticed the type annotations implemented here. I think that these annotations are pretty harmless, but just wanted to make sure the this is not in conflict with our decision on #2803?

skoudoro commented:
> I think that these annotations are pretty harmless, but just wanted to make sure the this is not in conflict with our decision on https://github.com/dipy/dipy/issues/2803?

I agree, they are harmless, but they are conflicting with our decision. I will remove them now and merge this PR.

Thank you for the feedback @arokem and thank you for this work @DelinteNicolas
